pkg/descheduler/node/node.go

```diff
@@ -96,76 +100,400 @@ func IsReady(node *v1.Node) bool {
+	ok, err := utils.PodMatchNodeSelector(pod, node)
+	if err != nil || !ok {
+		return false
+	}
+	// Check taints (we only care about NoSchedule and NoExecute taints)
+	ok = utils.TolerationsTolerateTaintsWithFilter(pod.Spec.Tolerations, node.Spec.Taints, func(taint *v1.Taint) bool {
+		return taint.Effect == v1.TaintEffectNoSchedule || taint.Effect == v1.TaintEffectNoExecute
+	})
+	if !ok {
+		return false
+	}
+	// Check if the pod can fit on a node based off it's requests
+	if !FitsRequest(ctx, client, pod, node) {
+		return false
+	}
+	// Check if node is schedulable
+	if IsNodeUnschedulable(node) {
+		return false
+	}
+
+	klog.V(2).InfoS("Pod can probably be scheduled on a different node once evicted", "pod", klog.KObj(pod), "node", klog.KObj(node))
+	return true
+}
 
+// PodFitsAnyOtherNode checks if the given pod will probably fit any of the given nodes, besides the node
+// the pod is already running on. The predicates used to determine if the pod will fit can be found in the NodeFit function.
+func PodFitsAnyOtherNode(ctx context.Context, client clientset.Interface, pod *v1.Pod, nodes []*v1.Node) bool {
 	for _, node := range nodes {
 		// Skip node pod is already on
 		if node.Name == pod.Spec.NodeName {
 			continue
 		}
-		// Check node selector and required affinity
-		ok, err := utils.PodMatchNodeSelector(pod, node)
-		if err != nil || !ok {
-			continue
-		}
-		// Check taints (we only care about NoSchedule and NoExecute taints)
-		ok = utils.TolerationsTolerateTaintsWithFilter(pod.Spec.Tolerations, node.Spec.Taints, func(taint *v1.Taint) bool {
-			return taint.Effect == v1.TaintEffectNoSchedule || taint.Effect == v1.TaintEffectNoExecute
-		})
-		if !ok {
-			continue
+
+		if NodeFit(ctx, client, pod, node) {
+			klog.V(2).InfoS("Pod fits on node", "pod", klog.KObj(pod), "node", klog.KObj(node))
+			return true
 		}
-		// Check if node is schedulable
-		if !IsNodeUnschedulable(node) {
-			klog.V(2).InfoS("Pod can possibly be scheduled on a different node", "pod", klog.KObj(pod), "node", klog.KObj(node))
+	}
+	return false
+}
+
+// PodFitsAnyNode checks if the given pod will probably fit any of the given nodes. The predicates used
+// to determine if the pod will fit can be found in the NodeFit function.
+func PodFitsAnyNode(ctx context.Context, client clientset.Interface, pod *v1.Pod, nodes []*v1.Node) bool {
+	for _, node := range nodes {
+		if NodeFit(ctx, client, pod, node) {
+			klog.V(2).InfoS("Pod fits on node", "pod", klog.KObj(pod), "node", klog.KObj(node))
 			return true
 		}
 	}
 	return false
 }
 
+// PodFitsCurrentNode checks if the given pod will probably fit onto the given node. The predicates used
+// to determine if the pod will fit can be found in the NodeFit function.
+func PodFitsCurrentNode(ctx context.Context, client clientset.Interface, pod *v1.Pod, node *v1.Node) bool {
+	if NodeFit(ctx, client, pod, node) {
+		klog.V(2).InfoS("Pod fits on node", "pod", klog.KObj(pod), "node", klog.KObj(node))
```

Comment:
Can all "fits" logs be changed to v4? It's very noisy in our cluster of thousands of pods.

Reply:
@a7i Great suggestions, I'm implementing these now. I'll report back here when everything is pushed.